File: src/main/java/ru/hollowhorizon/hc/common/registry/ModCapabilities.java
package ru.hollowhorizon.hc.common.registry;

import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.player.Player;
import net.minecraftforge.event.AttachCapabilitiesEvent;
import ru.hollowhorizon.hc.common.capabilities.HollowCapabilityStorageV2;

public class ModCapabilities {

    public static void init() {
        HollowCapabilityStorageV2.INSTANCE.registerAll();
    }

    public static void attachCapabilityToEntity(AttachCapabilitiesEvent<Entity> event) {
        if (event.getObject() instanceof Player) {
            //event.addCapability(new ResourceLocation(MODID, "player_capabilities"), new HollowPlayerProvider());
        }

    }


}

File: common/src/main/java/ru/hollowhorizon/hc/common/registry/ModCapabilities.kt
package ru.hollowhorizon.hc.common.registry

import net.minecraft.world.item.ItemStack
import ru.hollowhorizon.hc.api.ICapabilityDispatcher
import ru.hollowhorizon.hc.common.capabilities.CAPABILITIES
import ru.hollowhorizon.hc.common.events.SubscribeEvent
import ru.hollowhorizon.hc.common.events.capabilities.LoadCapabilitiesEvent

object ModCapabilities {
    @SubscribeEvent
    fun loadCapabilities(event: LoadCapabilitiesEvent) {
        CAPABILITIES
            .filter { it.key.isValid(event.provider) }
            .forEach {
                it.value.forEach {
                    event.addCapability(it())
                }
            }
    }

    private fun Class<*>.isValid(dispatcher: ICapabilityDispatcher): Boolean {
        return if (dispatcher is ItemStack) isInstance(dispatcher.item)
        else isInstance(dispatcher)
    }
}